sync-core/src/test/java/com/cloudant/sync/query/UnindexedMatcherTest.java

```diff
@@ -0,0 +1,394 @@
+        assertThat(UnindexedMatcher.compareEq(1.0, 1.0f), is(false));
+        assertThat(UnindexedMatcher.compareEq(1.0f, 1l), is(false));
+        assertThat(UnindexedMatcher.compareEq(1l, 1.0f), is(false));
+        assertThat(UnindexedMatcher.compareEq(1.0f, 1), is(false));
+        assertThat(UnindexedMatcher.compareEq(1, 1.0f), is(false));
+
+        assertThat(UnindexedMatcher.compareEq(1.0f, 1.1f), is(false));
+        assertThat(UnindexedMatcher.compareEq(1.1f, 1.0f), is(false));
+        assertThat(UnindexedMatcher.compareEq(1.1f, 1.0), is(false));
+        assertThat(UnindexedMatcher.compareEq(1.0, 1.1f), is(false));
+        assertThat(UnindexedMatcher.compareEq(1.1f, 1l), is(false));
+        assertThat(UnindexedMatcher.compareEq(1l, 1.1f), is(false));
+        assertThat(UnindexedMatcher.compareEq(1.1f, 1), is(false));
+        assertThat(UnindexedMatcher.compareEq(1, 1.1f), is(false));
+    }
+
+    @Test
+    public void compareEqStringToString() {
+        assertThat(UnindexedMatcher.compareEq("mike", "mike"), is(true));
+        assertThat(UnindexedMatcher.compareEq("mike", "Mike"), is(false));
+    }
+
+    @Test
+    public void compareEqStringToNumber() {
+        assertThat(UnindexedMatcher.compareEq("1", 1), is(false));
+        assertThat(UnindexedMatcher.compareEq(1, "1"), is(false));
+    }
+
+    @Test
+    public void compareEqDoubleToDouble() {
+        assertThat(UnindexedMatcher.compareEq(1.0, 1.0), is(true));
+        assertThat(UnindexedMatcher.compareEq(1.0, 1.00), is(true));
+        assertThat(UnindexedMatcher.compareEq(1.0, 1.1), is(false));
+        assertThat(UnindexedMatcher.compareEq(1.1, 1.0), is(false));
+    }
+
+    @Test
+    public void compareEqDoubleToLong() {
+        assertThat(UnindexedMatcher.compareEq(1.0, 1l), is(true));
+        assertThat(UnindexedMatcher.compareEq(1l, 1.0), is(true));
+        assertThat(UnindexedMatcher.compareEq(1.1, 1l), is(false));
+        assertThat(UnindexedMatcher.compareEq(1l, 1.1), is(false));
+    }
+
+    @Test
+    public void compareEqDoubleToInt() {
+        assertThat(UnindexedMatcher.compareEq(1.0, 1), is(true));
+        assertThat(UnindexedMatcher.compareEq(1, 1.0), is(true));
+        assertThat(UnindexedMatcher.compareEq(1.1, 1), is(false));
+        assertThat(UnindexedMatcher.compareEq(1, 1.1), is(false));
+    }
+
+    @Test
+    public void compareEqLongToLong() {
+        assertThat(UnindexedMatcher.compareEq(1l, 1l), is(true));
+        assertThat(UnindexedMatcher.compareEq(1l, 2l), is(false));
+        assertThat(UnindexedMatcher.compareEq(2l, 1l), is(false));
+    }
+
+    @Test
+    public void compareEqLongToInt() {
+        assertThat(UnindexedMatcher.compareEq(1l, 1), is(true));
+        assertThat(UnindexedMatcher.compareEq(1, 1l), is(true));
+        assertThat(UnindexedMatcher.compareEq(1l, 2), is(false));
+        assertThat(UnindexedMatcher.compareEq(2, 1l), is(false));
+    }
```

Comment:
Do you think you need to do the boxing manually instead of relying on autoboxing for the compareEq tests? Given that `compareEq` will never autobox because all its inputs will come from dictionaries where the boxing has already happened? I'd say you want to be passing Integers and Doubles in, and ensuring they work, and passing in Floats and ensuring they don't.

I'd also statically import compareEq if that's possible.